akka-stream/src/main/scala/akka/stream/impl/io/OutputStreamPublisher.scala

```diff
@@ -0,0 +1,125 @@
+
+  case class Write(bs: ByteString) extends OutputStreamPublisherInMessage
+
+  case class Flush() extends OutputStreamPublisherInMessage
+
+  case class Close() extends OutputStreamPublisherInMessage
+
+  trait OutputStreamPublisherOutMessage
+
+  case class Ok() extends OutputStreamPublisherOutMessage
+  case class Failure(msg: String) extends OutputStreamPublisherOutMessage
+
+}
+
+/** INTERNAL API */
+private[akka] class OutputStreamPublisher(bytesReadPromise: Promise[Long], bufSize: Int)
+  extends akka.stream.actor.ActorPublisher[ByteString]
+  with ActorLogging {
+
+  var availableChunks: Vector[ByteString] = Vector.empty[ByteString]
+
+  var bytesGotTotal = 0L
+
+  var notifyWhenSend: Option[ActorRef] = None
+  var notifyWhenSendAll: Option[ActorRef] = None
+
+  def receive = {
+    //Downstream messages
+    case ActorPublisherMessage.Request(elements) ⇒
+      signalDownstream()
+    case ActorPublisherMessage.Cancel ⇒
+      trySignalOnSend(signalErrorUpstream(_, "Stream is cancelled"))
+      context.stop(self)
+
+    //Upstream messages from OutputStreamAdapter
+    case OutputStreamPublisher.Write(bs) ⇒ onWrite(bs) //TODO what if it's been cancelled? need test
+    case OutputStreamPublisher.Flush()   ⇒ flush()
+    case OutputStreamPublisher.Close()   ⇒ flush(); onCompleteThenStop()
+  }
+
+  def signalDownstream(): Unit =
+    if (isActive) {
+      while (totalDemand > 0 && availableChunks.nonEmpty) {
+        val ready = availableChunks.head
+        availableChunks = availableChunks.tail
+        bytesGotTotal += ready.size
+        onNext(ready)
+        trySignalOnSend()
+      }
+      trySignalOnSendAll()
+    }
+
+  def trySignalOnSend(notify: (ActorRef) ⇒ Unit = signalOkUpstream(_)) = {
+    notifyWhenSend match {
+      case Some(sender) ⇒
+        notify(sender)
+        notifyWhenSend = None
+      case None ⇒
+    }
+  }
+
+  def trySignalOnSendAll() = {
+    if (availableChunks.isEmpty)
+      notifyWhenSendAll match {
+        case Some(sender) ⇒
+          signalOkUpstream(sender)
+          notifyWhenSendAll = None
+        case None ⇒
+      }
+  }
+  /*
+  Considering OutputStreamAdapter as upstream
+   */
+  def signalOkUpstream(sender: ActorRef) = sender ! OutputStreamPublisher.Ok
+
+  def signalErrorUpstream(actor: ActorRef, msg: String) = actor ! OutputStreamPublisher.Failure(msg)
+
+  override def postStop(): Unit = {
+    super.postStop()
+    bytesReadPromise.trySuccess(bytesGotTotal)
+  }
+
+  def isBufferFull = availableChunks.size == bufSize
+
+  def onWrite(byteString: ByteString) = {
+    if (!isCompleted && !isCanceled && !isErrorEmitted) {
+      availableChunks :+= byteString
+
+      if (isBufferFull)
+        notifyWhenSend = Some(sender)
+      else
+        sender ! OutputStreamPublisher.Ok
+
+      signalDownstream()
+    } else {
+      signalErrorUpstream(sender, "Stream is terminated, no writes are possible")
+    }
+  }
+
+  def flush() = {
+    notifyWhenSendAll = Some(sender)
```

Comment:
difference between `notifyWhenSendAll` and `notifyWhenSend` is a bit fuzzy, rename them or explain in impl docs?